app/routes.py

```diff
@@ -1,2 +1,179 @@
+def mark_task_complete(task_id):
+    task= Task.query.get(task_id)
+    if task is None:
+        return jsonify(task), 404
+
+    current_time = datetime.now()
+    task.completed_at = current_time
+
+    BOT_TOKEN = os.environ.get('BOT_TOKEN')
+    requests.post(
+        url="https://slack.com/api/chat.postMessage", 
+        headers={"Authorization": f"Bearer {BOT_TOKEN}"}, 
+        data={"channel": "task_notifications", "text": f"Someone just completed the task {task.title}"}
+    )
+
+    db.session.commit()
+
+    return jsonify({"task": task.task_dict()}), 200
+
+
+@tasks_bp.route("<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_task_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(task), 404
+    task.completed_at = None
+
+    db.session.commit()
+
+    return jsonify({"task": task.task_dict()}), 200
+
+
+@goals_bp.route("", methods=["GET", "POST"])
+def handle_goals():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body:
+            return jsonify({"details": "Invalid data"}), 400
+
+        new_goal = Goal(title = request_body["title"])
+
+        db.session.add(new_goal)
+        db.session.commit()
+
+        return jsonify({"goal": new_goal.goal_dict()}), 201
+
+    elif request.method == "GET":
+        request_body = request.get_json()
+        goals = Goal.query.all()
+        goal_response = [goal.goal_dict() for goal in goals]
+        return jsonify(goal_response), 200
+
+
+@goals_bp.route("/<goal_id>", methods = ["GET", "PUT", "DELETE"])
+def handle_one_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+    if goal is None:
+        return jsonify(goal), 404
+
+    request_body = request.get_json()
+    if request.method == "GET":
+        return jsonify({"goal": goal.goal_dict()}), 200
+
+    elif request.method == "PUT":
+        if "title" not in request_body:
+            return jsonify({"details": "Invalid data"}), 400
+
+        goal.title = request_body["title"]
+        db.session.commit()
+        return jsonify({"goal": goal.goal_dict()}), 200
+
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+        return jsonify({'details': f'Goal {goal.goal_id} "{goal.title}" successfully deleted'})
+
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST", "GET"])
+def handle_goal_and_task(goal_id):
+    goal = Goal.query.get(goal_id)
+    if goal is None:
+            return jsonify(goal), 404
+
+    if request.method == "GET":
+        return jsonify(goal.goal_and_task_dict()), 200
+
+    elif request.method=="POST":
+        request_body = request.get_json()
+
+        for task_id in request_body["task_ids"]:
+            task = Task.query.get(task_id)
+            if task is None:
+                return jsonify(None), 404
+            goal.tasks.append(task)
+            db.session.commit()
```

Comment:
Good job checking whether the each requested task id lookup is successful before trying to append it to the task list. However, you may want to wait to perform the appending and commiting until after you have validated _all_ the tasks, rather than task by task. With the task-by-task approach, we could end up partially modifying the tasks in a goal, but then fail partway through the list, leaving the goal partially modified. It's often desirable to ensure that either the entire change has been applied, or none of the change has.

Reply:
Appending to the tasks realtionship is absolutely a valid way to associate a task with a goal. Notice that it's not entirely clear from the project description or tests whether POSTing to the task list is _adding_ tasks to the goal (in addition to anything that was already there), or _replacing_ the tasks in the goal. Which perspective does your implementation take? How could we implement the other approach?